### ISSUE TITLE
cast throws internal error instead of contract failure

### ISSUE THREAD

**Issue by takikawa** [racket/typed-racket#368]
### What version of Racket are you using?

6.5 (also on HEAD)

### What program did you run?

```
Welcome to Racket v6.5.
-> (cast 3 (-> Integer Integer))
; procedure-arity: contract violation
;   expected: procedure?
;   given: 3
; [,bt for context]
```

### What should have happened?

A contract failure.

**Comment by takikawa** [comment]
Also reproducible with other contracts forms like `require/typed`:

```
Welcome to Racket v6.5.0.5.
-> (module a racket
     (define x 3) (provide x))
-> (require/typed 'a [x (-> Integer Integer)])
; procedure-arity: contract violation
;   expected: procedure?
;   given: 3
; [,bt for context]
```

**Comment by takikawa** [comment]
Looks like the problem is that `simple-result->` needs to do the first-order check in the late-neg projection as well before doing any of the chaperoning. Does that sound right?

(the first-order predicate isn't automatically called by the contract system before the late-neg projection AFAIK)

**Comment by samth** [comment]
I thought that it was, but @rfindler would be able to answer definitely.

**Comment by rfindler** [comment]
No, the projection is responsible for doing all checking.

**Comment by takikawa** [comment]
Fixed by a984281cdc3a343f1baedd940c3e7acb8eed15d8
